lec10-modelling.Rmd

```diff
@@ -1,4 +1,707 @@
+but it's often a good place to start to get an idea of what parameter range to check. 
+
+The idea behind least squares fitting is that we want to minimize the difference between
+our model's prediction and the actual data, and we do that by minimizing the sum of the
+**squares** of the **residuals**. Residuals (or **errors**) are the difference between
+the model and the data for each data point. The purpose of taking the square of each 
+residual is to take out the influence of the sign of the residual. 
+
+The sum of the squares of the residuals is just a number, and now we have a criteria we can use
+to determine which of two fits is better: whichever fit minimizes that number is the better fit.
+
+In practice, there are many ways to find the particular parameter that gives the best fit. 
+One way is to start at some parameter value, then start adjusting it by small amounts and
+checking if the fit gets better or worse. If it gets worse, go in a different direction. 
+If it gets better, keep going in that direction until it either starts getting worse again
+or the amount by which it gets better is very small. This type of algorithm is called
+**gradient descent**. 
+
+Another option is to try a range of parameter values and choose the one that gives the best
+fit out of that list. This is simpler to implement than the previous algorithm, but it's
+also computationally more costly --- it can take a long time. 
+
+We will do some examples of the second method today, what we'll call 'brute-force least squares'.  
+
+```{r}
+# Make a range of r parameter values to try
+r_vals <- seq(0.01, 0.3, by = 0.01)
+
+residuals_squared <- numeric(length(r_vals)) # empty vector to store results
+
+# loop through all the values, calculate what the data would look like with that value, and calculate the residuals
+
+for (i in seq(length(r_vals))) {
+  
+  r <- r_vals[i]
+  
+  # Use the known solution for the exponential model
+  # Use the first data point in the series as N0
+  predicted_N <- Ndata[1]*exp(r*times)  
+  
+  residuals <- predicted_N - Ndata
+  residuals_squared[i] <- sum(residuals^2)
+}
+```
+
+Let's plot the sum of residuals squared vs. $r$ to find which value of $r$ fits best:
+
+```{r}
+qplot(r_vals, residuals_squared)
+```
+
+We can see visually that the minimum is around $r = 0.2$, but to extract that number 
+from the list we can use the function `which`:
+
+```{r}
+best_fit <- which(residuals_squared == min(residuals_squared))
+
+r_fit <- r_vals[best_fit] 
+
+print(r_fit)
+```
+
+We got the 'correct' value of $r$, the one that we originally used to simulate the data.
+
+Finally, let's plot the fit against the original data:
+
+```{r}
+qplot(times,Ndata) +
+  geom_line(aes(x = times, y = Ndata[1] * exp(r_fit * times)))
+```
+
+You can play around with this simulated data and see what happens to the fit if you 
+change the amount of noise that gets added to the data.
+
+## Fitting models with two or more parameters
```

Comment:
It's great to see how what the students learnt in the intro R section comes together with the modelling part here and in the following sections!

Reply:
Should it be `###` based on the outline?